### PR TITLE
feat: bump to use node20 runtime

### LINKED ISSUE #366: github deprecated node16 actions, and node20 support is not yet in the roadmap for HashiCorps actions
### TL;DR
[Node16 has been on maintenance for over a year and end-of-life since 2023-09-11](https://github.com/nodejs/Release/#end-of-life-releases). On 2023-09-22 [node16 github actions have been deprecated](https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/) and to this date there's no work on adding support for node20 to HashiCorps actions.

Credit: the above was mostly copy-pasted from a similar issue for Google actions  https://github.com/google-github-actions/auth/issues/337#issue-1961228857

### PR DESCRIPTION
**Pull request by chenrui333**
Closes #366 

This is consistent with all the recent github action updates. (mostly in actions repos, like actions/checkout, actions/setup-node, etc.)

-------

Node12 was deleted from runner. Node20 was added to Actions Runner on v2.308.0.
Node16 has en end of life on 11 Sep 2023.

This PR updates the default runtime to node20, rather then node16

relates to https://github.com/actions/runner/pull/2732

### CODE REVIEW

**Comment by chenrui333** [comment]
needs a major version release after the PR merge

**Comment by chenrui333** [comment]
cc @bendbennett , this is ready to go. 🙏

**Comment by chenrui333** [comment]
cc @austinvalle

**Comment by chenrui333** [comment]
cc @austinvalle

**Comment by chenrui333** [comment]
I dont understand what is going on, can someone help take a look at this?

**Review by austinvalle — APPROVED** [review]
Hey there @chenrui333 👋🏻 , apologies on the delay here. This PR overall LGTM, we'll need a changelog but I can PR those in.

We haven't cut a release of this action in almost a year, so I want to double check some other things, but I plan to merge this PR and release a `v3.0.0` later this week.

File: .github/workflows/continuous-integration.yml [hunk 1]
```diff
@@ -9,7 +9,7 @@ on:
 jobs:
   check-dist:
     name: Check dist/ directory
-    uses: actions/reusable-workflows/.github/workflows/check-dist.yml@a8533f184b279cfc1b2dd6a96ed2f097ccf81189
+    uses: actions/reusable-workflows/.github/workflows/check-dist.yml@97a652b80035363df47baee5031ec8670b8878ac # 4.0.0
```
Comment (austinvalle) [inline .github/workflows/continuous-integration.yml:12]: Was moving to merge this and noticed this change, I don't see this commit SHA on `actions/reusable-workflows`, did you perhaps mean this?

```suggestion
    uses: actions/reusable-workflows/.github/workflows/check-dist.yml@967035ce963867fb956a309c9b67512314bc7c1f
```

I'm not sure why the GH workflow fail didn't show in the PR 🙃 
https://github.com/hashicorp/setup-terraform/actions/runs/6477487204/workflow

File: .github/workflows/continuous-integration.yml [hunk 2]
```diff
@@ -9,7 +9,7 @@ on:
...
   test:
     name: Test
-    uses: actions/reusable-workflows/.github/workflows/basic-validation.yml@a8533f184b279cfc1b2dd6a96ed2f097ccf81189
+    uses: actions/reusable-workflows/.github/workflows/basic-validation.yml@97a652b80035363df47baee5031ec8670b8878ac # 4.0.0
```
Comment (austinvalle) [inline .github/workflows/continuous-integration.yml:15]:
```suggestion
    uses: actions/reusable-workflows/.github/workflows/basic-validation.yml@967035ce963867fb956a309c9b67512314bc7c1f
```

**Review by austinvalle — CHANGES_REQUESTED** [review]

File: .github/workflows/continuous-integration.yml (hunk 1 above)
Comment (chenrui333) [reply .github/workflows/continuous-integration.yml:12]: not sure how I got it, let me revert these to latest main.

**Comment by chenrui333** [comment]
@austinvalle updated the action refs and rebased to the latest master, also add the changelog entry. Let me know if it works better now. Thanks!

**Comment by chenrui333** [comment]
@austinvalle all green 🎉

**Review by austinvalle — APPROVED** [review]
LGTM, I made a pushed a couple changes to the PR:
- Reverted some of the changes to `.github/workflows/continuous-integration.yml`, unforuntately our GH organization is limited to specific SHA commits of workflows that are approved. These are managed by a separate system so I just kept the node upgrades 👍🏻 
- Switched the changelog entry to be a `changie` entry

Thanks for your patience and contribution @chenrui333! I'm going to be releasing `v3.0.0` early next week once I merge a couple other PRs into `setup-terraform`

Closes #366

**Comment by chenrui333** [comment]
> LGTM, I pushed a couple changes to the PR:
> 
> * Reverted some of the changes to `.github/workflows/continuous-integration.yml`, unforuntately our GH organization is limited to specific SHA commits of workflows that are approved. These are managed by a separate system so I just kept the node upgrades 👍🏻
> * Switched the changelog entry to be a `changie` entry
> 
> Thanks for your patience and contribution @chenrui333! I'm going to be releasing `v3.0.0` early next week once I merge a couple other PRs into `setup-terraform`

Sounds good and thank you for getting this going!

File: .changes/unreleased/NOTES-20231027-085131.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+kind: NOTES
```
Comment (tom-reinders) [inline .changes/unreleased/NOTES-20231027-085131.yaml:1]: Should this not be a breaking change just like the last node update?
https://github.com/hashicorp/setup-terraform/blob/main/CHANGELOG.md?plain=1#L62-L64

Comment (chenrui333) [reply .changes/unreleased/NOTES-20231027-085131.yaml:1]: Besides this change would require newer runner versions, it should not break anything.

**Pull request merged**